### PR TITLE
fix(MeetingSdkAdapter): share button doesn't update the display

### PR DESCRIPTION
**Pull request by alinasuciu**
Fix the local share property to be updated  depending on the sdk event when the local share is stopped.
https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-233449

### CODE REVIEW

File: src/MeetingsSDKAdapter.js [hunk 1]
```diff
@@ -679,7 +679,7 @@ export default class MeetingsSDKAdapter extends MeetingsAdapter {
     }
 
     const enableSharingStream = async () => {
-      const [, localShare] = await sdkMeeting.getMediaStreams({sendShare: true});
+      const [localShare] = await sdkMeeting.getMediaStreams({sendShare: true});
```
Comment (lalli-flores) [inline src/MeetingsSDKAdapter.js:682]: doesn't `getMediaStreams` return an array with 2 media streams? one for video and one for screen share, if you get the first one, wouldn't that be the wrong one?

File: src/MeetingsSDKAdapter.js [hunk 2]
```diff
@@ -697,6 +697,11 @@ export default class MeetingsSDKAdapter extends MeetingsAdapter {
       // TODO (SPARK-148910): emit should be at the end of the function.
       // This is a workaround for a bug of async updateShare() call.
       // The async call may need ~30s to finish execution, or even cannot finish execution.
+      // The stopStream method should be called after updateShare async call
+      // The localShare propery of meeting should be updated after updateShare async call
+      this.stopStream(this.meetings[ID].localShare);
```
Comment (lalli-flores) [inline src/MeetingsSDKAdapter.js:702]: I'm not sure this is fixing the issue or could potentially cause more problems

File: src/MeetingsSDKAdapter.js [hunk 3]
```diff
@@ -41,8 +40,8 @@ import {
 const EVENT_MEDIA_READY = 'media:ready';
 const EVENT_MEDIA_STOPPED = 'media:stopped';
 const EVENT_STATE_CHANGE = 'meeting:stateChange';
-const EVENT_LOCAL_SHARE_STOP = 'meeting:stoppedSharingLocal';
-const EVENT_LOCAL_SHARE_START = 'meeting:startedSharingLocal';
+const EVENT_LOCAL_SHARE_STOPPED = 'meeting:stoppedSharingLocal';
+const EVENT_LOCAL_SHARE_STARTED = 'meeting:startedSharingLocal';
 const EVENT_REMOTE_SHARE_START = 'meeting:startedSharingRemote';
```
Comment (cipak) [inline src/MeetingsSDKAdapter.js:45]: seems that these event names also need to be updated. Maybe we can create a different task named "Fix SDK meeting event names used in adapter" just for that

File: src/MeetingsSDKAdapter.js [hunk 4]
```diff
@@ -679,36 +678,36 @@ export default class MeetingsSDKAdapter extends MeetingsAdapter {
     }
 
     const enableSharingStream = async () => {
-      const [, localShare] = await sdkMeeting.getMediaStreams({sendShare: true});
-
-      this.meetings[ID].localShare = localShare;
-
       // Due to SDK limitation around local media updates,
```
Comment (cipak) [inline src/MeetingsSDKAdapter.js:681]: maybe fix the comment: disable the button while the SDK is enabling sharing

File: src/MeetingsSDKAdapter.js [hunk 5]
```diff
@@ -679,36 +678,36 @@ export default class MeetingsSDKAdapter extends MeetingsAdapter {
...
 
+      const [, localShare] = await sdkMeeting.getMediaStreams({sendShare: true});
+
+      this.meetings[ID].localShare = localShare;
+
       await sdkMeeting.updateShare({stream: localShare, sendShare: true, receiveShare: true});
     };
 
     const disableSharingStream = async () => {
       // TODO (SPARK-148910): emit should be at the end of the function.
```
Comment (cipak) [inline src/MeetingsSDKAdapter.js:696]: "emit should be at the end..." - not anymore, you can remove this.

you can leave the comment about the call taking 30s, maybe something like:

// disable the control while the SDK is stopping sharing, which make take 30 seconds or more

File: src/MeetingsSDKAdapter.js [hunk 6]
```diff
@@ -679,36 +678,36 @@ export default class MeetingsSDKAdapter extends MeetingsAdapter {
...
       });
 
       await sdkMeeting.updateShare({
         sendShare: false,
         receiveShare: true,
       });
 
-      this.stopStream(this.meetings[ID].localShare);
-      this.meetings[ID].localShare = null;
+      // The rest of the cleanup is done in the handling of the STOP event emitted by sdkMeeting.updateShare
```
Comment (cipak) [inline src/MeetingsSDKAdapter.js:710]: STOP => EVENT_LOCAL_SHARE_STOPPED

File: src/MeetingsSDKAdapter.js [hunk 7]
```diff
@@ -806,21 +805,27 @@ export default class MeetingsSDKAdapter extends MeetingsAdapter {
...
-        // eslint-disable-next-line no-console
-        tap(() => console.log('EVENT_LOCAL_SHARE_STOP is triggered', this)),
-        map(() => inactiveShare),
-      );
+    const meetingWithMediaStoppedSharingLocalEvent$ = fromEvent(
+      sdkMeeting,
+      EVENT_LOCAL_SHARE_STOPPED,
+    ).pipe(
+      // eslint-disable-next-line no-console
+      tap(() => console.log('EVENT_LOCAL_SHARE_STOPPED is triggered', this)),
```
Comment (cipak) [inline src/MeetingsSDKAdapter.js:813]: maybe "was" instead of "is", as we are handling the event here, not triggering it

File: src/MeetingsSDKAdapter.js [hunk 8]
```diff
@@ -806,21 +805,27 @@ export default class MeetingsSDKAdapter extends MeetingsAdapter {
...
+      tap(() => {
+        this.meetings[ID].localShare = null;
+        this.stopStream(this.meetings[ID].localShare);
```
Comment (cipak) [inline src/MeetingsSDKAdapter.js:816]: won't work, the stream was just set to `null` in the line above.

File: src/MeetingsSDKAdapter.test.js [hunk 9]
```diff
@@ -639,6 +639,12 @@ describe('Meetings SDK Adapter', () => {
   });
 
   describe('shareControl()', () => {
+    beforeEach(() => {
+      meetingSDKAdapter.meetings[meetingID] = {
+        ...meeting,
+        localShare: {},
```
Comment (cipak) [inline src/MeetingsSDKAdapter.test.js:645]: why is this needed?

File: src/MeetingsSDKAdapter.test.js [hunk 10]
```diff
@@ -717,7 +723,6 @@ describe('Meetings SDK Adapter', () => {
       await meetingSDKAdapter.handleLocalShare(meetingID);
 
       expect(mockSDKMeeting.updateShare).toHaveBeenCalled();
-      expect(meetingSDKAdapter.meetings[meetingID].localShare).toBeNull();
```
Comment (cipak) [inline src/MeetingsSDKAdapter.test.js:720]: in order to leave this `expect` in place, we need to wait a little more. You can use https://www.npmjs.com/package/wait-for-expect

Comment (alinasuciu) [reply src/MeetingsSDKAdapter.test.js:720]: I don't think this should be here anymore since here we call the handleLocalShare method and we check if the disableSharingStream is triggered and this doesn't set the localShare property to null anymore.

File: src/MeetingsSDKAdapter.test.js (hunk 9 above)
Comment (alinasuciu) [reply src/MeetingsSDKAdapter.test.js:645]: otherwise it doesn't know about the localShare property

File: src/MeetingsSDKAdapter.js (hunk 1 above)
Comment (alinasuciu) [reply src/MeetingsSDKAdapter.js]: Honestly I didn't know which one in the right one since the naming doesn't correspond with the one from api documentation. I reverted the changes.

File: src/mockSdk.js [hunk 11]
```diff
@@ -14,7 +14,7 @@ export const mockSDKPerson = {
   orgId: 'orgID',
 };
 
-export const mockSDKMeeting = {
+export const createMockSDKMeeting = () => ({
```
Comment (cipak) [inline src/mockSdk.js:17]: jsdoc. And better use the "function createMockSDKMeeting" syntax instead of arrow.

**Review by cipak — APPROVED** [review]

File: src/MeetingsSDKAdapter.js [hunk 12]
```diff
@@ -679,36 +678,33 @@ export default class MeetingsSDKAdapter extends MeetingsAdapter {
...
+
       await sdkMeeting.updateShare({stream: localShare, sendShare: true, receiveShare: true});
     };
 
     const disableSharingStream = async () => {
-      // TODO (SPARK-148910): emit should be at the end of the function.
-      // This is a workaround for a bug of async updateShare() call.
-      // The async call may need ~30s to finish execution, or even cannot finish execution.
+      // Disable the control while the SDK is stopping sharing, which could take 30 seconds or more
+
```
Comment (lalli-flores) [inline src/MeetingsSDKAdapter.js:696]: You should remove this new line

File: src/MeetingsSDKAdapter.js [hunk 13]
```diff
@@ -806,21 +802,27 @@ export default class MeetingsSDKAdapter extends MeetingsAdapter {
...
-        // eslint-disable-next-line no-console
-        tap(() => console.log('EVENT_LOCAL_SHARE_STOP is triggered', this)),
-        map(() => inactiveShare),
-      );
+    const meetingWithMediaStoppedSharingLocalEvent$ = fromEvent(
+      sdkMeeting,
+      EVENT_LOCAL_SHARE_STOPPED,
+    ).pipe(
+      // eslint-disable-next-line no-console
+      tap(() => console.log('EVENT_LOCAL_SHARE_STOPPED was triggered', this)),
```
Comment (lalli-flores) [inline src/MeetingsSDKAdapter.js:810]: This should probably be informational
```suggestion
      tap(() => console.info('EVENT_LOCAL_SHARE_STOPPED was triggered', this)),
```

File: src/MeetingsSDKAdapter.js [hunk 14]
```diff
@@ -806,21 +802,27 @@ export default class MeetingsSDKAdapter extends MeetingsAdapter {
...
-        // eslint-disable-next-line no-console
-        tap(() => console.log('EVENT_LOCAL_SHARE_START is triggered')),
-        map(() => activeShare),
-      );
+    const meetingWithMediaStartedSharingLocalEvent$ = fromEvent(
+      sdkMeeting,
+      EVENT_LOCAL_SHARE_STARTED,
+    ).pipe(
+      // eslint-disable-next-line no-console
+      tap(() => console.log('EVENT_LOCAL_SHARE_STARTED was triggered', this)),
```
Comment (lalli-flores) [inline src/MeetingsSDKAdapter.js:823]:
```suggestion
      tap(() => console.info('EVENT_LOCAL_SHARE_STARTED was triggered', this)),
```

**Review by lalli-flores — COMMENTED** [review]
Minor comments but I think this implementation should work 👍🏼
Tests are not passing, however. Could you fix them?

File: src/mockSdk.js [hunk 15]
```diff
@@ -14,90 +14,97 @@ export const mockSDKPerson = {
...
-          participant: {
-            person: {
-              ...mockSDKPerson,
-              id: 'id',
+/**
+ * Creates a mock instance of the Webex SDK Meeting used in unit testing
+ *
+ * @returns {object} mockSDKMeeting Instance
+ */
+export function createMockSDKMeeting() {
```
Comment (cipak) [inline src/mockSdk.js:22]: in order to prevent all the indentation changes, can you please switch to the arrow function syntax?

```js
export const createMockSDKMeeting = () => ({
  id: 'meetingID',
  ...,
});
